File: mechanical/src/main/scala/resonantinduction/mechanical/energy/gearshaft/GearShaftNode.java
package resonantinduction.mechanical.energy.gearshaft;

import net.minecraft.tileentity.TileEntity;
import net.minecraftforge.common.ForgeDirection;
import calclavia.api.resonantinduction.IMechanicalNode;
import calclavia.lib.grid.INodeProvider;
import resonantinduction.mechanical.energy.gear.PartGear;
import resonantinduction.mechanical.energy.gear.PartGearShaft;
import resonantinduction.mechanical.energy.grid.MechanicalNode;

public class GearShaftNode extends MechanicalNode
{
    public GearShaftNode(PartGearShaft parent)
    {
        super(parent);
    }
    
    @Override
    public double getTorqueLoad()
    {
        // Decelerate the gear based on tier.
        switch (shaft().tier)
        {
            default:
                return 0.03;
            case 1:
                return 0.02;
            case 2:
                return 0.01;
        }
    }

    @Override
    public double getAngularVelocityLoad()
    {
        return 0;
    }

    @Override
    public void doRecache()
    {
        connections.clear();

        /** Check for internal connections, the FRONT and BACK. */
        for (int i = 0; i < 6; i++)
        {
            ForgeDirection checkDir = ForgeDirection.getOrientation(i);

            if (checkDir == shaft().placementSide || checkDir == shaft().placementSide.getOpposite())
            {
                MechanicalNode instance = ((INodeProvider) shaft().tile()).getNode(MechanicalNode.class, checkDir);

                if (instance != null && instance != this && instance.canConnect(checkDir.getOpposite(), this))
                {
                    connections.put(instance, checkDir);
                }
            }
        }

        /** Look for connections outside this block space, the relative FRONT and BACK */
        for (int i = 0; i < 6; i++)
        {
            ForgeDirection checkDir = ForgeDirection.getOrientation(i);

            if (!connections.containsValue(checkDir) && (checkDir == shaft().placementSide || checkDir == shaft().placementSide.getOpposite()))
            {
                TileEntity checkTile = new universalelectricity.api.vector.Vector3(shaft().tile()).translate(checkDir).getTileEntity(world());

                if (checkTile instanceof INodeProvider)
                {
                    MechanicalNode instance = ((INodeProvider) checkTile).getNode(MechanicalNode.class, checkDir.getOpposite());

                    // Only connect to shafts outside of this block space.
                    if (instance != null && instance != this && instance.parent instanceof PartGearShaft && instance.canConnect(checkDir.getOpposite(), this))
                    {
                        connections.put(instance, checkDir);
                    }
                }
            }
        }
    }

    @Override
    public boolean canConnect(ForgeDirection from, Object source)
    {
        if (source instanceof MechanicalNode)
        {
            if (((MechanicalNode) source).parent instanceof PartGear)
            {
                PartGear gear = (PartGear) ((MechanicalNode) source).parent;

                if (!(Math.abs(gear.placementSide.offsetX) == Math.abs(shaft().placementSide.offsetX) && Math.abs(gear.placementSide.offsetY) == Math.abs(shaft().placementSide.offsetY) && Math.abs(gear.placementSide.offsetZ) == Math.abs(shaft().placementSide.offsetZ)))
                {
                    return false;
                }
            }
        }

        return from == shaft().placementSide || from == shaft().placementSide.getOpposite();
    }

    @Override
    public boolean inverseRotation(ForgeDirection dir, IMechanicalNode with)
    {
        if (shaft().placementSide.offsetY != 0 || shaft().placementSide.offsetZ != 0)
        {
            return dir == shaft().placementSide.getOpposite();
        }

        return dir == shaft().placementSide;
    }

    public PartGearShaft shaft()
    {
        return (PartGearShaft) this.parent;
    }
}

File: src/main/scala/resonantinduction/mechanical/mech/gearshaft/GearShaftNode.scala
package resonantinduction.mechanical.mech.gearshaft

import net.minecraftforge.common.util.ForgeDirection
import resonant.api.grid.INodeProvider
import resonant.lib.transform.vector.Vector3
import resonant.lib.wrapper.ForgeDirectionWrapper._
import resonantinduction.core.interfaces.TMechanicalNode
import resonantinduction.mechanical.mech.gear.{NodeGear, PartGear}
import resonantinduction.mechanical.mech.grid.NodeMechanical

class GearShaftNode(parent: PartGearShaft) extends NodeMechanical(parent)
{
  override def getLoad: Double =
  {
    return shaft.tier match
    {
      case 0 => 0.03
      case 1 => 0.02
      case 2 => 0.01
    }
  }

  override def getAngularVelocityLoad: Double =
  {
    return shaft.tier match
    {
      case 0 => 0.03
      case 1 => 0.02
      case 2 => 0.01
    }
  }

  override def rebuild()
  {
    //Check only two possible sides for connections
    for (toDir <- Seq(shaft.placementSide, shaft.placementSide.getOpposite))
    {
      var found = false

      ///Check within this block for another gear plate that will move this shaft
      val otherNode = shaft.tile.asInstanceOf[INodeProvider].getNode(classOf[NodeMechanical], toDir)

      if (otherNode != null && otherNode != this && canConnect(otherNode, toDir) && otherNode.canConnect(this, toDir.getOpposite))
      {
        connect(otherNode, toDir)
        found = true
      }

      if (!found)
      {
        ///Check for other gear shafts outside this tile
        val checkTile = new Vector3(shaft.tile).add(toDir).getTileEntity(world)

        if (checkTile.isInstanceOf[INodeProvider])
        {
          val instance = (checkTile.asInstanceOf[INodeProvider]).getNode(classOf[NodeMechanical], toDir.getOpposite)

          if (instance != null && instance != this && instance.getParent.isInstanceOf[PartGearShaft] && instance.canConnect(this, toDir.getOpposite))
          {
            connect(instance, toDir)
          }
        }
      }
    }
  }

  override def canConnect[B](other: B, from: ForgeDirection): Boolean =
  {
    if (other.isInstanceOf[NodeMechanical])
    {
      if ((other.asInstanceOf[NodeMechanical]).getParent.isInstanceOf[PartGear])
      {
        val gear: PartGear = (other.asInstanceOf[NodeMechanical]).getParent.asInstanceOf[PartGear]
        if (!(Math.abs(gear.placementSide.offsetX) == Math.abs(shaft.placementSide.offsetX) && Math.abs(gear.placementSide.offsetY) == Math.abs(shaft.placementSide.offsetY) && Math.abs(gear.placementSide.offsetZ) == Math.abs(shaft.placementSide.offsetZ)))
        {
          return false
        }
      }
    }
    return from == shaft.placementSide || from == shaft.placementSide.getOpposite
  }

  override def inverseRotation(other: TMechanicalNode): Boolean = other.isInstanceOf[NodeGear] && other.asInstanceOf[NodeGear].parent.asInstanceOf[PartGear].placementSide.offset < Vector3.zero

  def shaft: PartGearShaft = getParent.asInstanceOf[PartGearShaft]
}